### PR TITLE
Custom domains verification

### LINKED ISSUE #80: CNAME support
**What**

GLBC does not allow custom hosts. Supporting custom hosts would provide a valuable feature and allow users to use a domain they own rather than the ugly subdomain provided by default

In order to provide CNAME support, we will need to do the following things:

- Verify the domain is owned via mechanism such as DNS Verification, similar to how letsencrpyt etc verify ownership
- Add the custom domain and the generated subdomain to the Ingress hosts and copy the paths for each host
- Allow the user to add a certificate via a secret for the custom domain

### PR DESCRIPTION
**Pull request by sergioifg94**
Closes #80 

## Verification steps
1. To mock the TXT verification, uncomment the following block, to make use the ConfigMap resolver:  https://github.com/Kuadrant/kcp-glbc/blob/3f8124e5c4ce84b9ca61226b01ab9b5de34748e8/cmd/kcp-glbc/main.go#L265-L269
2. Start KCP
    ```sh
    make local-setup
    ```

3. In another terminal, run the controller
    ```sh
    export KUBECONFIG=config/deploy/local/kcp.kubeconfig
    # If mocking the DNS management:
    # export AWS_DNS_PUBLIC_ZONE_ID=FAKE_ZONE_ID
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc
    ./bin/kcp-glbc --kubeconfig .kcp/admin.kubeconfig --context system:admin --enable-custom-hosts
    ```
4. If mocking the TXT verification, in another terminal, create the mock ConfigMap
    ```sh
    export KUBECONFIG=config/deploy/local/kcp.kubeconfig
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc
    kubectl apply -f samples/domainverfication/hosts.yaml
    ```
5. In another terminal, create the Ingress, DomainVerification and workload
    ```sh
    export KUBECONFIG=.kcp/admin.kubeconfig
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc-user
    kubectl apply -f samples/domainverification/example.yaml
    ```
6. Verify that the DomainVerification is reconciled, remains unverified and the token is set
7. Verify that the Ingress is reconciled and
    1. The custom host is replaced for the generated one
    2. The custom host rule is preserved in the `kuadrant.dev/custom-hosts.pending` annotation of the Ingress
8. Test the domain verification. Take the token from the `DomainVerification` resource and set it in the TXT entry
    > If mocking the TXT resolution, set it in the `hosts` ConfigMap
9. Verify that
    1. The DomainVerification resource is reconciled and marked as verified
    2. The Ingress resource is reconciled and the custom host is added back to the `spec` 

#### To verify the empty host scenario

1. Edit the Ingress, appending a rule with the empty host:
    ```yaml
    - host: ""
      http:
        paths:
        - backend:
            service:
              name: httpecho-both
              port:
                number: 80
          path: /mypath
          pathType: Prefix
    ```
2. Verify that the ingress is reconciled and the empty host is replaced by the generated host

### CODE REVIEW

**Comment by sergioifg94** [comment]
@maleck13 There's still a few small changes I haven't commited yet, but no harm in reviewing some of the current code

File: pkg/reconciler/domainverification/controller.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		c.Logger.Info("DomainVerfication was deleted", "key", key)
+		return nil
+	}
+
+	current := domainVerfication.(*v1.DomainVerification)
```
Comment (maleck13) [inline pkg/reconciler/domainverification/controller.go:73]: We should swap these around. Pass in the copy and then check against the original

Comment (maleck13) [reply pkg/reconciler/domainverification/controller.go:73]: @astefanutti Can you clarify the reason for this swap that was made in the ingress controller?

Comment (astefanutti) [reply pkg/reconciler/domainverification/controller.go:73]: The object passed to the `reconcile` method is usually mutated, which may leave a stale object in the work-queue if an error occurred, and broke subsequent informer event handlers and reconciliations / retries. Passing the disposable copy instead fixes that.

**Comment by maleck13** [comment]
@sergioifg94 I thought part of the design was that the DomainVerification would be created by the controller when a custom domain was discovered?
cc @philbrookes

**Comment by philbrookes** [comment]
https://github.com/Kuadrant/kcp-glbc/pull/323

**Pull request closed without merging**